File: leetcode/56. Merge Intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        res = []
        
        intervals.sort(key=lambda x:x[0])
        
        curr = intervals[0]
        
        for i in range(1, len(intervals)):
            interval = intervals[i]
            if interval[0] > curr[1]:
                res.append(curr)
                curr = interval
            else:
                curr[1] = max(curr[1], interval[1])
        
        res.append(curr)
        return res

File: leetcode/56. Merge Intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        std::sort(intervals.begin(), intervals.end(), [](auto& x, auto& y){
            return x[0] < y[0];
        });

        vector<vector<int>> res;
        if (intervals.empty()) {
            return res;
        }

        int start = intervals[0][0];
        int end = intervals[0][1];
        for (size_t i = 0; i < intervals.size(); ++i) {
            if (intervals[i][0] > end) {
                res.push_back({start, end});
                start = intervals[i][0];
                end = intervals[i][1];
            } else {
                end = std::max(end, intervals[i][1]);
            }
        }
        res.push_back({start, end});

        return res;
    }
};
